### PR TITLE
Fixed MousePointer

### PR DESCRIPTION
**Pull request by thalbern**
Overview
readded changes from https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/3982 and addressed issues raised:

New Changes
--- readded scroll and original input mappings
--- implemented spatial ptr 
--- simplyfied position calculation
--- added gameobject position setting to conform with posesynchronizer
--- removed spatial grip input mapping -> will be fixed in the manipulation handler by @luis-valverde-ms 

------------------------------------------------
Original Changes
--- mouse cursor not stuck anymore in center of screen
--- only one mouse cursor gameobject is created / controller won't be enabled more than once
--- screenspace cursor will be hidden when spatial cursor is used
--- fixed other controllers interferring with mouse cursor visibility (hands / camera)
--- removed spatial mouse cursor (retrieved the mouse cursor position and sent it back to the mouse cursor to set the position of the cursor to that same position again)
--- removed mouse wheel scrolling that modified mouse cursor position (not needed as spatial cursor is snapping to interactable objects)
--- added ray to world space position calc to enable object manipulation (bounding box)
--- added spatial grab to controller mapping to support dragging of objects
--- added mouse cursor speed adjustment setting

### CODE REVIEW

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs [hunk 1]
```diff
@@ -130,6 +126,16 @@ public void Update()
                         }
                     }
                 }
+
+                if (Interactions[i].InputType == DeviceInputType.Scroll)
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: What's the benefit of moving this block?

Comment (thalbern) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: no benefit. just happened while merging the old with the new commits. if that's an issue I'll move it back? let me know

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: Just easier to review without as much churn. No worries.

**Comment by Alexees** [comment]
would love to see this being approved as not being able to use the mouse slows down development quite a bit

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs [hunk 2]
```diff
@@ -65,32 +64,39 @@ public void Update()
...
+                    mouseDeltaRotation.y += UInput.GetAxis("Mouse X");
+                    IMixedRealityMousePointer mousePointer = (IMixedRealityMousePointer)InputSource.Pointers[0];
+                    if (mousePointer != null)
+                    {
+                        mouseDeltaRotation *= mousePointer.Speed;
+                    }
+
+                    MixedRealityPose controllerPose = MixedRealityPose.ZeroIdentity;
+                    controllerPose.Rotation = Quaternion.Euler(mouseDeltaRotation);
                     Interactions[i].PoseData = controllerPose;
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: >Interactions[i].PoseData = controllerPose [](start = 20, length = 41)

We seem to be storing just a rotation delta here. It should be a proper world space pose instead. The idea is to use that directly in the mouse pointer as ray cast origin.

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs [hunk 3]
```diff
@@ -65,32 +64,39 @@ public void Update()
...
             {
+                
+
                 if (Interactions[i].InputType == DeviceInputType.SpatialPointer)
                 {
+                    // add mouse delta as rotation
+                    var mouseDeltaRotation = Vector3.zero;
+                    mouseDeltaRotation.x += -UInput.GetAxis("Mouse Y");
+                    mouseDeltaRotation.y += UInput.GetAxis("Mouse X");
+                    IMixedRealityMousePointer mousePointer = (IMixedRealityMousePointer)InputSource.Pointers[0];
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: >IMixedRealityMousePointer mousePointer = (IMixedRealityMousePointer)InputSource.Pointers[0]; [](start = 20, length = 92)

Controllers shouldn't know about pointers. We can configure mouse speed either through a profile for the mouse device manager or by consuming just mouse deltas in the mouse pointer and applying it there.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs [hunk 4]
```diff
@@ -167,10 +171,23 @@ public override void OnInputChanged(InputEventData<Vector2> eventData)
...
-                    UpdateMousePosition(eventData.InputData.x, eventData.InputData.y);
+                    Vector3 mouseDeltaRotation = Vector3.zero;
+                    mouseDeltaRotation.x += eventData.InputData.x;
+                    mouseDeltaRotation.y += eventData.InputData.y;
+                    UpdateMouseRotation(mouseDeltaRotation * speed);
+                }
+            }
+        }
+
+        public override void OnInputChanged(InputEventData<MixedRealityPose> eventData)
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs]: >OnInputChanged [](start = 29, length = 14)

ControllerPoseSynchronizer is already doing this for us.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs [hunk 5]
```diff
@@ -75,15 +82,27 @@ public override IMixedRealityController Controller
         /// <inheritdoc />
         public override void OnPreSceneQuery()
         {
+            // screenspace to ray conversion
             transform.position = CameraCache.Main.transform.position;
 
-            Ray ray = new Ray(Position, Rotation * Vector3.forward);
+            Ray ray = new Ray(transform.position, transform.forward);
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs]: >Ray ray = new Ray(transform.position, transform.forward); [](start = 12, length = 57)

As far as I understand, the idea for pointers derived from controller pose synchronizer is that they will take the origin and direction from the pose reported by the controller. In that sense, here we should be using directly Position and Rotation * Vector3.forward for the ray.

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs [hunk 6]
```diff
@@ -75,15 +82,27 @@ public override IMixedRealityController Controller
...
+
+            // ray to worldspace conversion
+            gameObject.transform.position = transform.position + transform.forward * DefaultPointerExtent;
+        }
+
+        public override Vector3 Position
+        {       
+            get
+            {
+                return gameObject.transform.position;
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs]: >gameObject.transform.position [](start = 23, length = 29)

By overriding the base implementation here we are ignoring raycastOrigin. I think we could be better off having mouse pointer not derive from BaseControllerPointer / ControllerPoseSynchronizer.

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs (hunk 2 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: The problem I have here is that I don't have access to the previous mouse position - I shouldn't query the pointer so I can't provide the position here if we want to keep using the delta

File: Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs (hunk 4 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/MousePointer.cs]: the controllerPoseSynchronizer is not calling our method for calculating the rotation - I have to keep this method

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs (hunk 3 above)
Comment (thalbern) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: in the latest update there's a config profile for mouser input for setting the speed - that should solve this issue

File: Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs [hunk 7]
```diff
@@ -65,32 +65,40 @@ public void Update()
...
+                
+
                 if (Interactions[i].InputType == DeviceInputType.SpatialPointer)
                 {
+                    // add mouse delta as rotation
+                    var mouseDeltaRotation = Vector3.zero;
+                    mouseDeltaRotation.x += -UInput.GetAxis("Mouse Y");
+                    mouseDeltaRotation.y += UInput.GetAxis("Mouse X");
+                    
+                    var mouseInputProfile = MixedRealityToolkit.Instance?.GetService<MouseDeviceManager>()?.MouseInputProfile;
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: GetService is surprisingly expensive, it does a whole bunch of allocations and linear string search. Typical usage is to have an internal ref and lazy-initialize it once, e.g.: [https://github.com/microsoft/MixedRealityToolkit-Unity/blob/acee452b82ffbfe6f48c0c4b1086058496d94161/Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/SolverHandler.cs#L145](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/acee452b82ffbfe6f48c0c4b1086058496d94161/Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/SolverHandler.cs#L145)

Comment (thalbern) [reply Assets/MixedRealityToolkit/Providers/UnityInput/MouseController.cs]: caching profile now in controller and pointer

**Comment by thalbern** [comment]
added follow up task for removing posesynchronizer from mouse pointer https://github.com/microsoft/MixedRealityToolkit-Unity/issues/4251

**Pull request merged**